snuba/datasets/dataset.py

```diff
@@ -101,52 +144,37 @@ def process_condition(self, condition) -> Tuple[str, str, Any]:
         """
         return condition
 
-    def get_extensions(self) -> Mapping[str, QueryExtension]:
-        """
-        Returns the extensions for this dataset.
-        Every extension comes as an instance of QueryExtension.
-        The schema tells Snuba how to parse the query.
-        The processor actually does query processing for this
-        extension.
-        """
-        raise NotImplementedError("dataset does not support queries")
-
     def get_split_query_spec(self) -> Union[None, ColumnSplitSpec]:
         """
         Return the parameters to perform the column split of the query.
         """
         return None
 
-    def get_query_processors(self) -> Sequence[QueryProcessor]:
-        """
-        Returns a series of transformation functions (in the form of QueryProcessor objects)
-        that are applied to queries after parsing and before running them on Clickhouse.
-        These are applied in sequence in the same order as they are defined and are supposed
-        to be stateless.
-        """
-        return []
-
 
 class TimeSeriesDataset(Dataset):
     def __init__(
         self,
-        *args,
-        dataset_schemas: DatasetSchemas,
+        storages: Sequence[Storage],
+        query_plan_builder: StorageQueryPlanBuilder,
+        abstract_column_set: ColumnSet,
+        writable_storage: Optional[TableStorage],
```

Comment:
Isn't `writable_storage` one of the `storages` on line 157? Why does it need to get passed the same thing again? It's not entirely clear to me how many of each kind of storages per dataset are allowed.

Reply:
there has to be one writable storage (if we want to write). get_all_storages does not guarantee that there is only one.
Though I think the type of writable_storage should be a bit more restrictive.

Reply:
Shouldn't get_all_storages() guarantee 1 if we want to write or 0 if not, nothing else seems valid. And how many read storages can it contain?

Reply:
It can contain any number of read storages.
Now the issues in in using get_all_storages to fetch the single write storage are:
- Type: I am going to restrict the writable_storage to have its own type that is a subclass of TableStorage so no assertions anymore on whether the storage can write. Once you get one, you can write. get_all_storage would require an isinstance to see if we can write.
- having a get_writable_storage ensures that either you get 1 or 0 through type checking. Which is more reliable than the implicit assumption (for the callsite) that only one of the storages of get_all_storages will be writable. When we can enforce something through type checking it is better than if we have to make assumption to enforce invariants, since type checking guarantees can be automatically checked.
- if we assumed that only one storage in get_all_storages is writable, we would have to enforce this guarantee in two places: when we fill in the content of the sequence that returns all storages, and when we call get_all_storages (to be defensive, otherwise we would have to trust it).

Reply:
I guess I mean making the dataset more aware of which ones are read and write not. For example replacing `storages` with `read_storages` to skip passing the write storage twice and ensuring that an invalid number of write storages does not get passed into here. `get_all_storages()` can still return `read_storages + [write_storage]` (if there is a write storage).

Reply:
A writable storage is also a readable storage (same object), since it would extend Storage/TableStorage, so keeping two disjoint lists would actually impose a model that does not represent the reality, because then we would have to decide whether to add the writable storage to the readable ones or to the writable ones. 

The two disjoint lists can be implemented only if we ReadStorage and WriteStorage had not common parent class. I need to think over this to see if it is possible.

Reply:
What about this. Let's frame the problem this way:
requirements:
- DDL related operations (bootstrap, migrate), just need a list of storages to get schemas from
- executing a query needs a list of storage we can query, but not only, they need to know exactly which storage is for, see discover, so it would not use a get_all_storages method or a get_read_storages.
- consuming ideally would be fed with a dataset_identifier+storage_identifier and not a dataset identifier so we could have multiple  writer per dataset (discover could be the writer for both transactions and events). so it needs a feature that returns a writable storage given an id. Potentially this could even return directly a TableWriter instead of the whole storage. Now there is only one and there is no id, so we can replace get_writer_by_id with get_writer.

Thus I don't see the need of a get_readable_storages method returning a list. of storages. 
We need a get_all_storages for DDL and just the writers for the consumers. So I do see the need to identify the writable storages but I don't see a good case for exposing which ones are readers, who needs to know which storages are readers is the StorageQueryPlanBuilder which already is provided only the storages it should use and each storage has a specific role.

Does this make sense?

Reply:
That helps, I thought we would also have write only storages and cannot be read from.

> consuming ideally would be fed with a dataset_identifier+storage_identifier and not a dataset identifier

Why does consuming need to know about the dataset identifier anymore? Doesn't this responsibility move into the storage class now? Also does this also imply the same storage can't be shared between datasets?

Reply:
potentially it may not, but then we need to ensure the storage id is unique across the system instead of within a dataset. Which I guess we can do

Reply:
> * Type: I am going to restrict the writable_storage to have its own type that is a subclass of TableStorage so no assertions anymore on whether the storage can write. Once you get one, you can write. get_all_storage would require an isinstance to see if we can write.

👍 I noted this in a comment elsewhere but good to see we're on the same page.

> When we can enforce something through type checking it is better than if we have to make assumption to enforce invariants, since type checking guarantees can be automatically checked.

💯 

> I thought we would also be having write only storages that cannot be read from

I think this will actually be true at some point (see other comment … somewhere this review), but I think that we should avoid introducing this complexity until we have to — as long as we keep that in mind while making (and reviewing) these changes it hopefully will be a more straightforward change at that point than trying to also tackle that right now.

Reply:
split writable and readable storage